SRVKE-652: Container source docs (#34707)
openshift/openshift-docs: modules/serverless-containersource-custom-sources.adoc

[id="serverless-containersource-custom-sources_{context}"]
= Creating custom event sources by using a container source

You can use a container source to create and manage a container for your custom event source image.

In order to implement a custom event source by using a container source, you must first create a container image of your event source, and then create a container source that specifies the correct configuration, including the container image URI.

== Guidelines for creating a container image

* The container image can be developed using any language, and can be built and published with any tool you like.

> * The main process of the container image must accept parameters from arguments and environment variables.

Author: lol this was a suggestion from QE. Previously it was `main` method, but @cardil pointed out that since folks can use many different languages to write their source, it's better to refer to it in a more generic way. I would say it's OK as is and will probably make sense to developers, unless there's some other suggestion about how to word it.

> * The container image must start with a `main` method, which accepts parameters from arguments and environment variables.

Reviewer: Not every language has a `main` method. So, to make it more general, I'd say:
```suggestion
* The container image main process needs to accept parameters from arguments and environment variables.
```

* Two environment variables are injected by the container source controller; `K_SINK` and `K_CE_OVERRIDES`. These variables are resolved from the `sink` and `ceOverrides` spec, respectively.

* Event messages are sent to the sink URI specified in the `K_SINK` environment variable. The event message can be in any format, however using the link:https://cloudevents.io/[`CloudEvent`] spec is recommended.

> .Example heartbeats container image

Author: Great suggestions, thank you!

[source,go]
----
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"os"
	"strconv"
	"time"

	duckv1 "knative.dev/pkg/apis/duck/v1"

	cloudevents "github.com/cloudevents/sdk-go/v2"

> 	"github.com/kelseyhightower/envconfig"

Author: LMAO I didn't even notice this, good catch 😄 @cardil @openshift/team-knative-eventing can someone please provide a clean example for this instead

> 	"github.com/kelseyhightower/envconfig"

Reviewer 2: I mean, he's a huge figure in open source and k8s so I'm not convinced this _isn't_ a good working example. 😀

> 	"github.com/kelseyhightower/envconfig"

Reviewer: His envconfig package is widely used. What's the problem with that?

> 	"github.com/kelseyhightower/envconfig"

Author: Fair enough lol just thought it was odd to have someone's name in it but that's OK with me if it's OK with you
